filters/auth/grantcookie.go

```diff
@@ -0,0 +1,134 @@
+package auth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/zalando/skipper/secrets"
+	"golang.org/x/oauth2"
+)
+
+type cookie struct {
+	AccessToken  string    `json:"access_token"`
+	RefreshToken string    `json:"refresh_token"`
+	Expiry       time.Time `json:"expiry,omitempty"`
+	RefreshAfter time.Time `json:"refresh_after,omitempty"`
+}
+
+func refreshAfter(expiry time.Time) time.Time {
+	now := time.Now()
+	if now.After(expiry) {
+		return now
+	}
+
+	d := expiry.Sub(now)
+	d /= 10
+	if d < time.Minute {
+		d = time.Minute
+	}
+
+	return now.Add(d)
+}
+
+func decodeCookie(cookieHeader string, config OAuthConfig) (c *cookie, err error) {
+	var eb []byte
+	if eb, err = base64.StdEncoding.DecodeString(cookieHeader); err != nil {
+		return
+	}
+
+	var encryption secrets.Encryption
+	if encryption, err = config.Secrets.GetEncrypter(secretsRefreshInternal, config.SecretFile); err != nil {
+		return
+	}
+
+	var b []byte
+	if b, err = encryption.Decrypt(eb); err != nil {
+		return
+	}
+
+	err = json.Unmarshal(b, &c)
+	return
+}
+
+func (c *cookie) isAccessTokenExpired() bool {
+	now := time.Now()
+	return now.After(c.Expiry)
+}
+
+// extractCookie removes and returns the OAuth Grant token cookie from a HTTP request.
+// The function supports multiple cookies with the same name and returns
+// the best match (the one that decodes properly).
+// The client may send multiple cookies if a parent domain has set a
+// cookie of the same name.
+// The grant token cookie is extracted so it does not get exposed to untrusted downstream
+// services.
+func extractCookie(request *http.Request, config OAuthConfig) (cookie *cookie, err error) {
+	old := request.Cookies()
+	new := make([]*http.Cookie, 0, len(old))
+
+	for i, c := range old {
+		if c.Name == config.TokenCookieName {
+			cookie, _ = decodeCookie(c.Value, config)
+			if cookie != nil {
+				new = append(new, old[i+1:]...)
+				break
+			}
+		}
+		new = append(new, c)
+	}
+
+	if cookie != nil {
+		request.Header.Del("Cookie")
+		for _, c := range new {
+			request.AddCookie(c)
+		}
+		return cookie, nil
+	}
+	return nil, http.ErrNoCookie
+}
+
+func CreateCookie(config OAuthConfig, host string, t *oauth2.Token) (*http.Cookie, error) {
```

Comment:
we don't need to export this function. If it is currently exported only because the grant_test.go uses `grant_test` as the package name, then we can just create another additional test file, using `grant` as the package name for those tests that directly call the `createCookie()` function.

Reply:
You are right in that we only export his function so that we can use it in the test. It looks to me like this can't easily be changed though since the tests in `grant_test` require quite a bit of scaffolding (e.g. `grantQueryWithCookie`) and a lot of code would have to be moved as well. Or maybe I'm missing something here?
Looking at this made me realise though: The two new test files (`grantflowstate_test.go` and `grant_test.go`) are actually the only ones that have `auth_test` as a package. All others re-use `auth`. Would changing these tests to also live in the `auth` package be an alternative here?

Reply:
Yes, @k-nut you can create a separate test file with `auth` package and add your tests there that need access to `createCookie()`

Reply:
I added 1f8301226dfebda06e560f97e7ebcc8569ff0ad9 which introduces a couple of new helpers functions for the test to get test cookies (inspired by the one that we had already). Would this be acceptable?

Reply:
wait, that commit is broken. Fixing.

Reply:
Ok, this is commit 17b9590 now after amending

Reply:
yes that should be fine IIRC